util/containers/Dockerfile

```diff
@@ -0,0 +1,86 @@
+# syntax=docker/dockerfile:1.4
+
+ARG NODE_VERSION
+ARG RUST_VERSION
+
+
+FROM --platform=${BUILDPLATFORM} docker.io/node:${NODE_VERSION}-alpine AS frontend-build
+
+WORKDIR /build
+COPY frontend/package.json frontend/package-lock.json /build/
+RUN npm i
+
+COPY frontend .
+RUN npm ci
+RUN npx relay-compiler
+RUN npx webpack --progress --mode=production
+
+
+FROM docker.io/rust:${RUST_VERSION}-alpine AS backend-build
+
+ENV CFLAGS    "-fdebug-prefix-map=/root/.cargo/registry/src/github.com-1ecc6299db9ec823/=__dep__"
+ENV RUSTFLAGS "--remap-path-prefix=/build/backend=<src> --remap-path-prefix=/root/.cargo/registry/src/github.com-1ecc6299db9ec823/=<dep>"
+
+WORKDIR /build/backend
+
+RUN apk add --no-cache \
+  make \
+  musl-dev \
+  perl
+
+ARG RUST_TARGET
+RUN rustup target add "${RUST_TARGET}"
+
+COPY backend/Cargo.toml backend/Cargo.lock /build/backend/
+COPY <<EOF /build/backend/src/main.rs
+fn main() {}
+EOF
+RUN cargo build --target "${RUST_TARGET}" --release
```

Comment:
Again this: can you explain why this is necessary/useful?

Reply:
Same principle as with npm. This will cache a build of all Rust dependencies. If versions do not change, these will not be rebuilt again.

There is a relatively new alternative that could be used: instead of caching in layers, one can mount a caching directory during the build where output between builds can be reused. Would you prefer that?

Reply:
See main-thread comment below for more comments.